Add support for namespaced maps

iarenaza opened the pull request:
parseclj already added support for namespaced maps in 2018 (in commit b40670a56147214f0486763529897cb688a09692).

As Alex Miller said in https://github.com/edn-format/edn/issues/78

> Clojure introduced namespace map syntax in Clojure 1.9. The Clojure
> edn reader was also updated to support the non-autoresolved parts of
> namespace map syntax
>
> The edn spec should be updated to a new version that includes the
> namespace map syntax such as #:foo{:bar 1} (syntax alternative for
> {:foo/bar 1}).

So it's pretty clear that the intention is for edn spec to support namespaced maps (the reference implementation in Clojure already does!).

In addition to the use case mentioned in issue #16, not supporting this seems to cause trouble in CIDER when integrating with shadow-cljs (https://github.com/clojure-emacs/cider/issues/3437

Fixes: #16

iarenaza committed:
Add support for namespaced maps

parseclj already added support for namespaced maps in 2018 (in commit
b40670a56147214f0486763529897cb688a09692).

As Alex Miller said in https://github.com/edn-format/edn/issues/78

> Clojure introduced namespace map syntax in Clojure 1.9. The Clojure
> edn reader was also updated to support the non-autoresolved parts of
> namespace map syntax (edn doesn't do anything autoresolved like
> ::foo or #::foo{}).
>
> The edn spec should be updated to a new version that includes the
> namespace map syntax such as #:foo{:bar 1} (syntax alternative for
> {:foo/bar 1}).

So it's pretty clear that the intention is for edn spec to support
namespaced maps (the reference implementation in Clojure already
does!).

In addition to the use case mentioned in issue #16, not supporting
this seems to cause trouble in CIDER when integrating with
shadow-cljs (https://github.com/clojure-emacs/cider/issues/3437)

Fixes: #16

vemv commented on test/parseedn-test-data.el:
```diff
@@ -243,6 +243,36 @@
 
+       "prefixed-map"
+       (a-list
+        :source "#:foo.bar{:baz 1 :other.ns.prefix/qux 2}"
```
A few variations that you might want to include into the test suite:

* `#:foo.bar {` (i.e. space before the bracket)
* `#::{:bar 2}` (expands to the current ns)
* `#::foo{:bar 2}` (expands to an aliased ns)

iarenaza replied:
While the first variation makes total sense, second and third don't (to me). As far as I know, auto-resolved keywords are not supported in EDN. I omitted a part of that detail in my quote in the commit message, but Alex Miller said https://github.com/edn-format/edn/issues/78: 

> The Clojure edn reader was also updated to support the non-autoresolved parts of namespace map syntax **(edn doesn't do anything autoresolved like ::foo or #::foo{})**.

(emphasis mine). Which IMHO, makes total sense. EDN is not Clojure and doesn't have the concept of the current namespace. The following code (using Clojure 1.9.0) show that the syntax is not supported by the reference implementation in Clojure:

```
user=> (require '[clojure.edn :as edn])
nil
user=> (edn/read-string "#::{:bar 2}")
RuntimeException Invalid token: :  clojure.lang.Util.runtimeException (Util.java:221)

user=> (edn/read-string "#::foo{:bar 2}")
RuntimeException Namespaced map must specify a valid namespace: :foo  clojure.lang.EdnReader$NamespaceMapReader.invoke (EdnReader.java:495)

user=> (edn/read-string "{::foo/bar 2}")
RuntimeException Invalid token: ::foo/bar  clojure.lang.Util.runtimeException (Util.java:221)

user=> (edn/read-string "{::bar 2}")
RuntimeException Invalid token: ::bar  clojure.lang.Util.runtimeException (Util.java:221)

user=>
```

vemv replied:
Got it, thanks!

True, parseedn != parseclj.

vemv commented on test/parseedn-test-data.el:
```diff
@@ -243,6 +243,36 @@
+       (a-list
+        :source "#:foo.bar{:baz 1 :other.ns.prefix/qux 2}"
+        :edn (list (a-hash-table :foo.bar/baz 1 :other.ns.prefix/qux 2))
+        :ast '((:node-type . :root)
```
You can also assert that the `:ast` of `"#:foo.bar{:baz}"` is equal to that of `{:foo.bar/baz}"`

iarenaza replied:
Well, that is not the case. `parseclj` produces different a different `:ast` for each case. Because it parses the map namespace (what `parseclj` calles the "map prefix") as a separate element of the AST. That is, it doesn't perform the `#:foo.bar{:baz 1}` -> `{:foo.bar/baz 1}` transformation as part of the parsing and AST generation. It produces two AST elements, one for the map prefix, and a second one for the map itself (note: I'm using `parseclj-parse-clojure` in the examples to make them self-contained, but the generated AST is exactly the same when parsing the examples as EDN instead as of Clojure):


```elisp
*** Welcome to IELM ***  Type (describe-mode) or press C-h m for help.
ELISP> (require 'parseclj)
parseclj
ELISP> (parseclj-parse-clojure "#::foo.bar{:baz 1}")
((:node-type . :root)
 (:position . 1)
 (:children
   ((:map-prefix
      (:token-type . :map-prefix)
      (:form . "#::foo.bar")
      (:pos . 1))
    (:node-type . :map)
    (:position . 11)
    (:children
      ((:node-type . :keyword)
       (:position . 12)
       (:form . ":baz")
       (:value . :baz))
      ((:node-type . :number)
       (:position . 17)
       (:form . "1")
       (:value . 1))))))

ELISP> 
```
You can see that the AST for `{:foo.bar/baz 1}` doesn't have the `:map-prefix` element at all:

```elisp
ELISP> (parseclj-parse-clojure "{:foo.bar/baz 1}")
((:node-type . :root)
 (:position . 1)
 (:children
   ((:node-type . :map)
    (:position . 1)
    (:children
      ((:node-type . :keyword)
       (:position . 2)
       (:form . ":foo.bar/baz")
       (:value . :foo.bar/baz))
      ((:node-type . :number)
       (:position . 15)
       (:form . "1")
       (:value . 1))))))

ELISP>
```

That is why we need to do the "assembling" (transformation) of the pieces in `parseedn-reduce-branch` (which is 95% of  what this patch does :smile:).

vemv replied:
Got it!

Still, would it be possible to assert a resemblance? e.g. `(= x (dissoc y :foo))`

What we gain is a higher-level test.

Otherwise, it can be easy for the `:ast`s to become sort of a blob without a particular meaning. In other words, a test that asserts that `(= x <large object>)` doesn't necessarily assert a specific property, so later, if the test changes, it's hard to see if it's still valid.

iarenaza replied:
First of all, thanks a lot for your feedback!

Unfortunately, I don't I understand what you mean by "assert a resemblance" and "gain a higher-level test". I don't know if you are referring to the `:ast` value when you talk about those two things, or the `:edn` and `:source` values.

I am not an expert in the `parseedn` package testing strategy, But from what I can understand from `define-parseedn-read-tests` and `define-parseedn-roundtrip-tests` (in `parsedn-test.el`), the `:ast` key defined in all the data definitions present in `parseedn-test-data.el`, that key is not used at all. So I would say that (at least at the moment), that key value is more informational that anything else.

Also, `parseedn-read` never sees the full value of the `:ast` as defined in `parseedn-test-data.el`. In fact, it doesn't know about it at all. The only functions that see anything related to the `:ast` value are `parseedn-reduce-leaf` and `parseedn-reduce-branch`. But they are called back from `parseclj` when needed, to "reduce" (transform the current AST node definition into the associated Emacs Lisp data type). In the former case to reduce the individual nodes of the `:ast` tree that are considered "leaf tokens" by `parseclj`. In the latter to reduce the sub-tree of the `:ast` tree that is considered a "branch" by `parseclj` (e.g., maps, lists, vectors, etc).

So could you please clarify what you'd like to test?

vemv replied:
I'll try this out locally, and if I find a meaningful test that can be added, will just post it here

iarenaza replied:
Merged!

vemv replied:
As you say, `parseedn-read never sees the full value of the :ast as defined in parseedn-test-data.el`.

So, let's remove all `:ast` values. They very likely are remnants of when parseedn and parseclj were the same library.

I have verified that tests keep passing.

If you want a patch:

    curl https://gist.githubusercontent.com/vemv/bbd7da50ffa8e82b685d905dfc276f4e/raw/43f411ac4b944d5e61643926b01b0cb09dfdd901/gistfile1.txt | git apply

iarenaza replied:
All `:ast` keys removed.

vemv reviewed (commented):
Very nice to see this! Kudos.

I cannot meaningfully review the implementation, however I spotted some opportunity in the tests.

iarenaza committed:
Remove all `:ast` values

parseedn-read never sees the full value of the `:ast` keys as defined
in `parseedn-test-data.el`. In fact, it doesn't know about it at
all. The only functions that see anything related to the `:ast` values
are `parseedn-reduce-leaf` and `parseedn-reduce-branch`. But they are
called back from `parseclj` when needed, to "reduce" (transform the
current AST node definition into the associated Emacs Lisp data type).

As commented in https://github.com/clojure-emacs/parseedn/pull/17#discussion_r1321814637,
they very likely are remnants of when parseedn and parseclj were the
same library.

vemv commented on parseedn.el:
```diff
@@ -113,10 +113,24 @@ on available options."
         ((eq :set token-type) (list 'edn-set children))
         ((eq :lbrace token-type) (let* ((kvs (seq-partition children 2))
-                                        (hash-map (make-hash-table :test 'equal :size (length kvs))))
+                                        (hash-map (make-hash-table :test 'equal :size (length kvs)))
```
Comparing the 'before' and 'after' of `defun parseedn-reduce-branch`, I sense that now it mixes levels of abstraction.

Before it looked more concise and higher-level.

The simplest fix would be to extract most of the logic to a separate defun. Although possibly this might mean something else (e.g. this could be fixed in parseclj?) - analysis much welcome.

vemv reviewed (commented):
I intent to merge this soon - was simply leaving some headroom in case someone else also wanted to review.

One last suggestion.

iarenaza committed:
Extract the logic of building the map to separate defuns

Before adding the code to support prefixed maps, the branch reduction
function looked more concise and higher-level. After adding the support
it seemed like it mixed levels of abstraction.

Extracting the nitty-gritty details of how to build the maps
(especially in the prefixed map case) should help in keeping the
levels of abstraction separated.
